syncer/syncer.go

```diff
@@ -0,0 +1,235 @@
+package syncer
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/SkynetLabs/blocker/blocker"
+	"github.com/SkynetLabs/blocker/database"
+	"github.com/SkynetLabs/blocker/skyd"
+	"github.com/sirupsen/logrus"
+	"gitlab.com/NebulousLabs/errors"
+	"go.sia.tech/siad/build"
+	"go.sia.tech/siad/crypto"
+)
+
+var (
+	// syncInterval defines the amount of time between syncs of external
+	// portal's blocklists, which can be defined in the environment using the
+	// key BLOCKER_SYNC_LIST
+	syncInterval = build.Select(
+		build.Var{
+			Dev:      time.Minute,
+			Testing:  time.Second,
+			Standard: 4 * time.Hour,
+		},
+	).(time.Duration)
+)
+
+type (
+	// Syncer periodically scans the blocklist of a set of portals which are
+	// configured by the user and adds the missing hashes to the blocklist of
+	// the local skyd. Alongside syncing with other portal's blocklists, the
+	// syncer will also ensure the local skyd's blocklist has all of the links
+	// which are contained in the database this syncer is connected to, ensuring
+	// all hashes are on its blocklist.
+	Syncer struct {
+		staticCtx        context.Context
+		staticDB         *database.DB
+		staticBlocker    *blocker.Blocker
+		staticLogger     *logrus.Logger
+		staticPortalURLs []string
+		staticSkydAPI    skyd.API
+	}
+
+	// lookupTable is a helper struct that defines a hash map
+	lookupTable map[crypto.Hash]struct{}
+)
+
+// New returns a new Syncer with the given parameters.
+func New(ctx context.Context, blocker *blocker.Blocker, skydAPI skyd.API, db *database.DB, portalURLs []string, logger *logrus.Logger) (*Syncer, error) {
+	if ctx == nil {
+		return nil, errors.New("invalid context provided")
+	}
+	if blocker == nil {
+		return nil, errors.New("invalid blocker provided")
+	}
+	if db == nil {
+		return nil, errors.New("invalid DB provided")
+	}
+	if logger == nil {
+		return nil, errors.New("invalid logger provided")
+	}
+	if skydAPI == nil {
+		return nil, errors.New("invalid Skyd API provided")
+	}
+	s := &Syncer{
+		staticBlocker:    blocker,
+		staticCtx:        ctx,
+		staticDB:         db,
+		staticLogger:     logger,
+		staticPortalURLs: portalURLs,
+		staticSkydAPI:    skydAPI,
+	}
+	return s, nil
+}
+
+// Start launches a background task that periodically syncs the blocklists of
+// the preconfigured portals with the blocklist of the local skyd instance.
+func (s *Syncer) Start() {
+	// convenience variables
+	logger := s.staticLogger
+
+	// sync the local mongo database with skyd
+	err := s.syncDatabaseWithSkyd()
```

Comment:
Why are we only ever syncing the db once? Is there no chance it gets updated in the meantime? Or do we have another module in the blocker which updates skyd from the db? In that case the architecture would be a bit chaotic imo.

Reply:
restructured to be less chaotic, this was actually made redundant by the changes I was planning to make already, when writing the syncer I just didn't realize this was not necessary any longer